crates/chia-consensus/src/multiprocess_validation.rs

```diff
@@ -0,0 +1,243 @@
+use crate::consensus_constants::ConsensusConstants;
+use std::sync::{Arc, Mutex};
+use std::thread;
+use crate::gen::condition_tools::pkm_pairs;
+use crate::gen::flags::{
+    AGG_SIG_ARGS, ALLOW_BACKREFS, ENABLE_MESSAGE_CONDITIONS, ENABLE_SOFTFORK_CONDITION,
+    NO_RELATIVE_CONDITIONS_ON_EPHEMERAL,
+};
+use crate::gen::owned_conditions::OwnedSpendBundleConditions;
+use crate::gen::solution_generator::solution_generator;
+use crate::gen::validation_error::ErrorCode;
+use crate::generator_types::BlockGenerator;
+use crate::npc_result::get_name_puzzle_conditions;
+use chia_bls::aggregate_verify;
+use chia_bls::BlsCache;
+use chia_bls::PublicKey;
+use chia_protocol::Coin;
+use chia_protocol::Program;
+use chia_protocol::SpendBundle;
+use chia_traits::Streamable;
+use clvmr::{ENABLE_BLS_OPS_OUTSIDE_GUARD, ENABLE_FIXED_DIV};
+use std::time::{Duration, Instant};
+
+// currently in multiprocess_validation.py
+// called via blockchain.py from full_node.py when a full node wants to add a block or batch of blocks
+fn pre_validate_blocks_multiprocessing() {}
+
+// currently in multiprocess_validation.py
+// called in threads from pre_validate_blocks_multiprocessing
+fn batch_pre_validate_blocks() {}
+
+// currently in mempool_manager.py
+// called in full_node.py when adding a transaction
+pub fn pre_validate_spendbundle(
+    new_spend: SpendBundle,
+    max_cost: u64,
+    constants: ConsensusConstants,
+    peak_height: u32,
+    syncing: bool,
+    cache: Arc<Mutex<BlsCache>>,
+) -> Result<OwnedSpendBundleConditions, ErrorCode> {
+    if new_spend.coin_spends.is_empty() {
+        Err(ErrorCode::InvalidSpendBundle)
+    } else {
+        let (result, duration) = validate_clvm_and_signature(
+            &new_spend,
+            max_cost,
+            constants,
+            peak_height,
+            syncing,
+            cache,
+        )?;
+        Ok(result)
+    }
+}
+
+// currently in mempool_manager.py
+// called in threads from pre_validate_spend_bundle()
+// returns (error, cached_results, new_cache_entries, duration)
+fn validate_clvm_and_signature(
+    spend_bundle: &SpendBundle,
+    max_cost: u64,
+    constants: ConsensusConstants,
+    height: u32,
+    syncing: bool,
+    cache: Arc<Mutex<BlsCache>>,
+) -> Result<(OwnedSpendBundleConditions, Duration), ErrorCode> {
+    let start_time = Instant::now();
+    let additional_data = constants.agg_sig_me_additional_data;
+    let program: BlockGenerator = simple_solution_generator(&spend_bundle)?;
+    let npcresult = get_name_puzzle_conditions(program, max_cost, true, height, constants);
```

Comment:
I think you could simplify this quite a bit by, instead of implementing the full `get_name_puzzle_conditions()`, take a look at `run_block_generator2()` ([here](https://github.com/Chia-Network/chia_rs/blob/main/crates/chia-consensus/src/gen/run_block_generator.rs#L113)). Now imagine you never have any block references (which we don't in the mempool) and also imagine that you already have all the spends as a list of puzzle-reveals and solutions (as we do in `spend_bundle`).

You don't actually need to generate a "simple solution" nor do you need to serialize and deserialize it.

You just need this section, but simplified, since you don't need to parse it out of the CLVM structure.

https://github.com/Chia-Network/chia_rs/blob/main/crates/chia-consensus/src/gen/run_block_generator.rs#L156-L192